### PR TITLE
main: Add "help" command

### PR DESCRIPTION
**Pull request by carlescufi**
Adds a `west help <command>` command that simply invokes the corresponding `-h`.

### CODE REVIEW

File: src/west/main.py [hunk 1]
```diff
@@ -503,16 +509,17 @@ def main(argv=None):
...
                 sys.executable, sys.argv[0], argv))
-        if sys.platform == "win32":
-            # On Windows Python generates .exe executables
-            os.execv(sys.argv[0], argv)
-        else:
-            os.execv(sys.executable, [sys.executable] + [sys.argv[0]] + argv)
+        # Treat the Python script as an executable
+        proc = Popen([sys.argv[0]] + argv)
+        proc.communicate()
+        sys.exit(proc.returncode or 0)
```
Comment (carlescufi) [inline src/west/main.py]: I had `or errno.EIO` here but apparently for whatever reason `proc.returncode` is sometimes `None` and the unit tests failed. Any ideas?

File: src/west/commands/common.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2019, Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''West common commands'''
```
Comment (mbolivar) [inline src/west/commands/common.py]: Would you mind just putting this in a `help.py` file for now, to make it easy for newcomers to find?

File: src/west/commands/common.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+from west.commands import WestCommand
+from west.util import WestReload
+
+class Help(WestCommand):
+    def __init__(self):
+        super().__init__(
+            'help',
+            'print west command help text',
+            _wrap('''
```
Comment (mbolivar) [inline src/west/commands/common.py]: When the help text is only a single line, you should be able to just use a single quoted string. We use triple quotes (and the related helper functions for controlling whitespace) only when there are multiple lines.

File: src/west/commands/common.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            Print help text on any command.
+            '''))
+
+    def do_add_parser(self, parser_adder):
+        return _add_parser(
+            parser_adder, self,
+            _arg('command', metavar='cmd', nargs='?',
+                 help='''Command to get help on''')
```
Comment (mbolivar) [inline src/west/commands/common.py]: Ditto here

File: src/west/commands/common.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    # later passed in parents= to add_parser()
+
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument(*args, **kwargs)
+    return parser
+
+
+# Arguments shared between more than one command
+
+def _add_parser(parser_adder, cmd, *extra_args, **kwargs):
```
Comment (mbolivar) [inline src/west/commands/common.py]: Is it worth moving _arg, _add_parser, and _wrap to helpers importable from `west.command` instead of keeping separate copies in project.py and this new file?

File: src/west/main.py [hunk 6]
```diff
@@ -48,15 +55,14 @@
 # Built-in commands in this West. For compatibility with monorepo
 # installations of West within the Zephyr tree, we only expose the
 # project commands if this is a multirepo installation.
```
Comment (mbolivar) [inline src/west/main.py]: Mind updating this comment while you are here?

File: src/west/main.py [hunk 7]
```diff
@@ -508,11 +508,8 @@ def main(argv=None):
         # latest version, with the same arguments that we were given.
         log.dbg("sys.executable:\"{}\" sys.argv[0]:\"{}\" argv:\"{}\"".format(
```
Comment (mbolivar) [inline src/west/main.py]: I guess this debug log needs an update too. Maybe just log `sys.argv[0]` and `argv`.

File: src/west/main.py [hunk 8]
```diff
@@ -26,9 +26,16 @@
 from west.commands import external_commands, \
     CommandError, CommandContextError, ExtensionCommandError
 from west.commands.project import List, ManifestCommand, Diff, Status, \
-    SelfUpdate, ForAll, WestUpdated, PostInit, Update
+    SelfUpdate, ForAll, PostInit, Update
+from west.commands.common import Help
 from west.manifest import Manifest, MalformedConfig
-from west.util import quote_sh_list
+from west.util import quote_sh_list, WestReload
```
Comment (mbolivar) [inline src/west/main.py]: I kind of feel like WestReload belongs in `west.commands`, since commands should be the only things that can raise the exception. Not married to the idea, though. What do you think?

File: src/west/main.py [hunk 9]
```diff
@@ -517,7 +517,9 @@ def main(argv=None):
         log.dbg("sys.executable:\"{}\" sys.argv[0]:\"{}\" argv:\"{}\"".format(
                 sys.executable, sys.argv[0], argv))
         # Treat the Python script as an executable
-        os.execv(sys.argv[0], argv)
+        proc = Popen([sys.argv[0]] + argv)
```
Comment (mbolivar) [inline src/west/main.py]: Does this break control-C in a gdb started by `west debug`? Attempts to run west as a subprocess generally do, because the SIGINT gets delivered to the parent process, which exits, instead of the subprocess west, which is temporarily ignoring it so GDB can handle it.

File: src/west/main.py (hunk 1 above)
Comment (mbolivar) [reply src/west/main.py]: That makes no sense; communicate should wait until the process is done. A None returncode, per documentation, means the subprocess isn't done executing yet.

But please see above about whether this approach is workable.

File: src/west/commands/common.py (hunk 2 above)
Comment (carlescufi) [reply src/west/commands/common.py]: Agreed

File: src/west/commands/common.py (hunk 3 above)
Comment (carlescufi) [reply src/west/commands/common.py]: ok!

File: src/west/commands/common.py (hunk 5 above)
Comment (carlescufi) [reply src/west/commands/common.py]: Yes, I think this makes sense. I will take a look.

File: src/west/main.py (hunk 6 above)
Comment (carlescufi) [reply src/west/main.py]: Will do

File: src/west/main.py (hunk 9 above)
Comment (carlescufi) [reply src/west/main.py]: It might have,  but `west debug` will never raise the exception. Only  commands that update repos and the help command can raise it. So no interactive commands raise the exception.

File: src/west/main.py (hunk 1 above)
Comment (carlescufi) [reply src/west/main.py]: It's strange yes. I will  debug this further and update.

File: src/west/main.py (hunk 9 above)
Comment (carlescufi) [reply src/west/main.py]: To add to this, we will remove the `selfupdate` when we move everything to pypi and also the current `execv` is actually broken on Windows, there's a race and sometimes you can end up with having to press enter to return to the shell. Finally, I think it's more likely that control-C will work reliably with `Popen` than with `execv`, at least on Windows

File: src/west/main.py (hunk 8 above)
Comment (carlescufi) [reply src/west/main.py]: I am indifferent in this case. I guess it could be raised by main itself in the future but I am fine with either

File: src/west/main.py (hunk 9 above)
Comment (carlescufi) [reply src/west/main.py]: Re: control-C you can see that actually both child and parent get the signal [here](https://stackoverflow.com/a/19807536/337201)

File: src/west/main.py (hunk 8 above)
Comment (mbolivar) [reply src/west/main.py]: Let's keep it here for now until we're ready to refactor to commit to a real API

File: src/west/main.py (hunk 1 above)
Comment (carlescufi) [reply src/west/main.py]: Doh, `0 or x` yields x... :|

File: src/west/commands/common.py (hunk 2 above)
Comment (carlescufi) [reply src/west/commands/common.py]: Done

File: src/west/commands/common.py (hunk 3 above)
Comment (carlescufi) [reply src/west/commands/common.py]: Done

File: src/west/commands/common.py (hunk 4 above)
Comment (carlescufi) [reply src/west/commands/common.py]: Done

File: src/west/commands/common.py (hunk 5 above)
Comment (carlescufi) [reply src/west/commands/common.py]: moved

File: src/west/main.py (hunk 6 above)
Comment (carlescufi) [reply src/west/main.py]: updated

File: src/west/main.py (hunk 7 above)
Comment (carlescufi) [reply src/west/main.py]: updated

File: tests/west/commands/test_commands.py [hunk 10]
```diff
@@ -190,6 +190,22 @@ def west_update_tmpdir(west_init_tmpdir):
...
+    assert(h1out == h2out)
+
+    cmds = ['help', 'list', 'manifest', 'diff', 'status', 'update',
+            'selfupdate', 'forall']
+
+    for c in cmds:
+        h1out = cmd('help {}'.format(c))
+        h2out = cmd('{} -h'.format(c))
+        assert(h1out == h2out)
+
```
Comment (tejlmand) [inline tests/west/commands/test_commands.py]: I think there should be an additional blank line here.

File: tests/west/commands/test_commands.py [hunk 11]
```diff
@@ -282,7 +298,6 @@ def test_manifest_freeze(west_update_tmpdir):
     for eline_re, aline in zip(expected_res, actual):
         assert re.match(eline_re, aline) is not None, (aline, eline_re)
 
-
```
Comment (tejlmand) [inline tests/west/commands/test_commands.py]: Why is this line removed ?

File: src/west/commands/project.py [hunk 12]
```diff
@@ -857,7 +823,7 @@ def _update_west(rebase, keep_descendants):
             # Signal self-update, which will cause a restart. This is a bit
             # nicer than doing the restart here, as callers will have a
             # chance to flush file buffers, etc.
-            raise WestUpdated()
+            raise WestReload()
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply src/west/commands/project.py]: Yes but the issue here is that with `west help` we have *not* updated west, we just want to reload it with different parameters. In the same way that you signal a `RuntimeError` you can signal a `WestReload`.

File: src/west/commands/command.py [hunk 13]
```diff
@@ -129,8 +184,6 @@ def do_run(self, args, unknown):
                         this object, this parameter is an empty sequence.
                         Otherwise, it is an iterable containing all unknown
                         arguments present on the command line.'''
-
-
 class WestExtCommandSpec:
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply src/west/commands/command.py]: Too much whitespace :)

File: tests/west/commands/test_commands.py (hunk 11 above)
Comment (carlescufi) [reply tests/west/commands/test_commands.py]: Too much whitespace ;)

File: src/west/commands/project.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply src/west/commands/project.py]: Then I understand the need for a `rename`, but I still think `WestReload` is bad, cause we haven't reloaded (yet).
Not sure what a better name would be though.

File: src/west/commands/help.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    def do_add_parser(self, parser_adder):
+        return self.add_parser_helper(
+            parser_adder,
+            self.add_arg('command', metavar='cmd', nargs='?',
+                         help='Command to get help on')
+        )
+
+    def do_run(self, args, user_args):
+        a = [args.command, '-h']
+        wr = WestReload(argv=[x for x in a if x is not None])
```
Comment (tejlmand) [inline src/west/commands/help.py]: Testing this out, I noticed that if an invalid command name is given, then user gets the default `west` usage.
```
$ west help unknown_cmd
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
west: error: argument <command>: invalid choice: 'unknown_cmd' (choose from 'list', 'manifest', 'diff', 'status', 'update', 'selfupdate', 'forall', 'post-init', 'help', 'build', 'flash', 'd
```
But in this case, `help` which is a valid command was given.

In contrast , if too many args a given to `west help` you'll get a `west help` error message.
```
$ west help list list
usage: west help [-h] [cmd]
west help: error: unexpected arguments: ['list']
```

Unfortunately, i'm not sure this is so easy to fix.

File: src/west/commands/help.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        raise wr
```
Comment (tejlmand) [inline src/west/commands/help.py]: Just doing some more thinking here.
Are there any reasons why we explicitly need to relaunch `west` ?
It should be possible to just call into `west.main` as we do with bootstrapper and wrapped commands.

The reason why `west update` has to relaunch, is because the `west` source code have changed when the SHAs differs, and thus we need to run rest of the upgrade process using the new `west` code.

File: tests/west/commands/test_commands.py (hunk 11 above)
Comment (tejlmand) [reply tests/west/commands/test_commands.py]: I thought you were using a black background on your screen.

File: src/west/commands/help.py (hunk 15 above)
Comment (tejlmand) [reply src/west/commands/help.py]: Not relaunching `west` could also make it easier to fix this https://github.com/zephyrproject-rtos/west/pull/204#discussion_r258152013

Comment (tejlmand) [reply src/west/commands/help.py]: This seems to work the same, and avoids the overhead of an exception and relaunching:
```
        import west.main
        west.main.main(argv=[x for x in a if x is not None])
```
instead of:
```
       wr = WestReload(argv=[x for x in a if x is not None])
       raise wr
```
Disclaimer, not fully tested at this point.

Comment (tejlmand) [reply src/west/commands/help.py]: Note: `west help init` will need special handling with suggested approach, as `west init -h` is solely handled in the bootstrapper today.

File: src/west/commands/project.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply src/west/commands/project.py]: As commented here: https://github.com/zephyrproject-rtos/west/pull/204#discussion_r258153115

I actually think we should only relaunch `west` if there really is a need, such as when `west` itself has a changed SHA due to an `selfupdate`

File: src/west/commands/help.py (hunk 15 above)
Comment (carlescufi) [reply src/west/commands/help.py]: I am very reluctant to do that because there's a whole bunch of things that are initialized and running inside the same Python VM instance makes me uncomfortable. Also it is very ugly to call back into what just called you IMHO?
@mbolivar thoughts?

File: src/west/commands/help.py (hunk 14 above)
Comment (carlescufi) [reply src/west/commands/help.py]: Yeah, I noticed that too. We can fix this later but it's not going to be easy because of the way we use ArgParse. It's `parse_known_ags` that deals with that automagically for us and displays the help text.

**Comment by tejlmand** [comment]
Should `west help` be moved to the bootstrapper ?

Doing `west help init` fails when not in project:
```
$ west help init
Error: "/home" is not in a west installation.
Things to try:
 - Set ZEPHYR_BASE to a zephyr repository path in a west installation.
 - Run "west init" to set up an installation here.
 - Run "west init -h" for additional information.
```

`west init -h` of course works.

**Comment by carlescufi** [comment]
>Should `west help` be moved to the bootstrapper ?

That's an interesting idea. What I don't like about it is that it's going to be a temporary solution anyway (we are going to go full pypi soon) and that it would not reuse all of the ArgParse/WestCommand architecture.

File: src/west/commands/help.py (hunk 15 above)
Comment (tejlmand) [reply src/west/commands/help.py]: @carlescufi in general I agree with you, but with the `west help` we know exactly where we are.
And seeing the `west help <cmd>` as a wrapper for `west <cmd> -h` I think that is nicer in this particular case.

I think it might be possible to fetch directly into `<command>.arg_parser` to get the help text we need.
That could make it an even nicer solution, than calling `west.main.main`
@mbolivar You have bit more experience with arg parser, would should approach seem feasible ?

My main point is that we don't need to relaunch.

**Comment by tejlmand** [comment]
> > Should `west help` be moved to the bootstrapper ?
> 
> That's an interesting idea. What I don't like about it is that it's going to be a temporary solution anyway (we are going to go full pypi soon) and that it would not reuse all of the ArgParse/WestCommand architecture.

Yep, just wanted to pint out the fact that `west help init` is not working for bootstrapper.
If that is acceptable, I'm fine with that :)

File: tests/west/commands/test_commands.py (hunk 11 above)
Comment (tejlmand) [reply tests/west/commands/test_commands.py]: Could you please put the blank line back ?
I believe we are following pep8 for python code:
https://www.python.org/dev/peps/pep-0008/

> Blank Lines
>
> Surround top-level function and class definitions with two blank lines.

**Comment by mbolivar** [comment]
> take a look at the two missing blank lines.

Let me explain what's going on here.

There are 2 blank lines between function and class definitions in several places in the code because indeed, that is pep8 style and flake8's default configuration complains when they are missing.

However (as shown here and in other PRs where @carlesc has snuck in whitespace changes removing them into other patches 😉) it seemed to me that most people didn't like that rule, so I disabled that check in our flake8 configuration:

https://github.com/zephyrproject-rtos/west/blob/master/tox.ini#L10

So in general I am ok with just one blank line. I do agree that in general whitespace "fixes" like this should be kept in their own patches though since they tend to clutter diffs and distract attention from the behavioral changes, also as shown here.

**Comment by tejlmand** [comment]
>  it seemed to me that most people didn't like that rule, so I disabled that check in our flake8 configuration:

Personally, i'm fine with single line and i'm fine with double lines.
What I dislike is mixing, i.e. some places with one line and other with double lines.
So if going for single blank line, I think all code should be adopted to that.

**Comment by carlescufi** [comment]
@mbolivar addressed all your comments now.

**Comment by mbolivar** [comment]
> So if going for single blank line, I think all code should be adopted to that.

I agree but haven't wanted to speak up about this yet, we have a lot of refactoring to do anyway.

Can we agree to hold off on further whitespace changes until we are ready to convert the entire code base?

File: src/west/commands/help.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2019, Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''West common commands'''
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply src/west/commands/help.py]: This file doesn't exist anymore

File: src/west/commands/command.py [hunk 17]
```diff
@@ -129,8 +184,6 @@ def do_run(self, args, unknown):
                         this object, this parameter is an empty sequence.
                         Otherwise, it is an iterable containing all unknown
                         arguments present on the command line.'''
-
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply src/west/commands/command.py]: There's actually 2 right now

File: src/west/commands/command.py [hunk 18]
```diff
@@ -100,6 +102,56 @@ def add_parser(self, parser_adder):
...
+        '''
+        parser = argparse.ArgumentParser(add_help=False)
+        parser.add_argument(*args, **kwargs)
+        return parser
+
+    @staticmethod
+    def wrap_help(s):
+        ''' Wraps help texts for commands.
+
+        Some of them have variable length (due to _MANIFEST_REV_BRANCH), so
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply src/west/commands/command.py]: This is no longer applicable, `command.py` has not changed anymore
